### PR TITLE
fix: slotted content of slotted custom element within if not visible

### PR DESCRIPTION
**Pull request by jodarove**
after being toggled 2 times.

## Details
When passing slotted content to a custom element which is also slotted and wrapped in a `if:true={foo}` directive and toggle the if multiple times, the passed slot content stops being visible after the second toggle. You can see a repro here: https://developer.salesforce.com/docs/component-library/tools/playground/mFcLmzSYw/37/edit

The main issue is that the `vnode` for the slotted custom element is being reused, and during slot allocation the first time, `vnode.children` is being emptied. 

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

### CODE REVIEW

**Comment by gonzalocordero** [comment]
@mrigdon FYI

File: packages/integration-karma/test/template/directive-if/index.spec.js [hunk 1]
```diff
@@ -126,6 +127,28 @@ describe('if:true directive', () => {
...
+        const elm = createElement('x-multiple-slot', { is: MultipleSlot });
+        document.body.appendChild(elm);
+        elm.shadowRoot.querySelector('.textToggle').click();
+
+        return Promise.resolve()
+            .then(() => {
+                expect(elm.shadowRoot.textContent).toContain('text in multiple level slot');
+                // hide the slot
+                elm.shadowRoot.querySelector('.textToggle').click();
+                return Promise.resolve();
```
Comment (ekashida) [inline packages/integration-karma/test/template/directive-if/index.spec.js]: Do we need this line?

File: packages/integration-karma/test/template/directive-if/index.spec.js [hunk 2]
```diff
@@ -126,6 +127,28 @@ describe('if:true directive', () => {
...
+            })
+            .then(() => {
+                // show the slot
+                elm.shadowRoot.querySelector('.textToggle').click();
+                return Promise.resolve();
```
Comment (ekashida) [inline packages/integration-karma/test/template/directive-if/index.spec.js]: Do we need this line?

File: packages/integration-karma/test/template/directive-if/index.spec.js [hunk 3]
```diff
@@ -126,6 +127,28 @@ describe('if:true directive', () => {
                 });
         });
     }
+
+    it('should render slot content multiple levels deep when toggled 2 times', () => {
```
Comment (ekashida) [inline packages/integration-karma/test/template/directive-if/index.spec.js]:
```suggestion
    it('should continue rendering content for nested slots after multiple rehydrations', () => {
```

File: packages/integration-karma/test/template/directive-if/x/multipleSlotLevel1/multipleSlotLevel1.html [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<template>
+    <template if:true={visible}>
+        <section data-visible={visible}>
+            <slot data-visible={visible}></slot>
```
Comment (ekashida) [inline packages/integration-karma/test/template/directive-if/x/multipleSlotLevel1/multipleSlotLevel1.html]:
```suggestion
        <section>
            <slot></slot>
```

File: packages/@lwc/engine/src/3rdparty/snabbdom/types.ts [hunk 5]
```diff
@@ -54,6 +54,7 @@ export interface VCustomElement extends VElement {
     mode: 'closed' | 'open';
     ctor: any;
     clonedElement?: undefined;
+    aChildren?: VNodes;
```
Comment (pmdartus) [inline packages/@lwc/engine/src/3rdparty/snabbdom/types.ts]: Can you add a comment here describing what this new property used for?

File: packages/integration-karma/test/template/directive-if/index.spec.js [hunk 6]
```diff
@@ -126,6 +127,28 @@ describe('if:true directive', () => {
...
+
+        return Promise.resolve()
+            .then(() => {
+                expect(elm.shadowRoot.textContent).toContain('text in multiple level slot');
+                // hide the slot
+                elm.shadowRoot.querySelector('.textToggle').click();
+                return Promise.resolve();
+            })
+            .then(() => {
+                // show the slot
```
Comment (pmdartus) [inline packages/integration-karma/test/template/directive-if/index.spec.js]: Assert the `textContent` is now `''`.

Comment (jodarove) [reply packages/integration-karma/test/template/directive-if/index.spec.js]: wow, this one was tricky in native shadow thanks! 👍

File: packages/@lwc/engine/src/3rdparty/snabbdom/types.ts [hunk 7]
```diff
@@ -54,6 +54,11 @@ export interface VCustomElement extends VElement {
     mode: 'closed' | 'open';
     ctor: any;
     clonedElement?: undefined;
+    // This VCustomElement may be reused when is part of the slot content in another slotted VCustomElement.
```
Comment (caridy) [inline packages/@lwc/engine/src/3rdparty/snabbdom/types.ts]: @jodarove this is the text that will show up when you roll over the key in VSCode, make it smaller, something like, "copy of the last list of children collection"

**Pull request merged**